File: 801→900/802. Find Eventual Safe States/Solution.cpp
class Solution {
    public:
        // Perform DFS to detect cycles starting from a given node
        bool dfs(int node, vector<vector<int>>& adj, vector<bool>& visit,
                 vector<bool>& inStack, vector<bool>& safe) {
            // If the node is already in the stack, we have a cycle.
            if (inStack[node]) {
                return true;
            }
            if (visit[node]) {
                // If the node is visited and not in the stack, it means we've already checked it.
                return false;
            }
            // Mark the current node as visited and part of the current recursion stack.
            visit[node] = true;
            inStack[node] = true;
            
            // Check all neighbors for cycles
            for (auto neighbor : adj[node]) {
                if (dfs(neighbor, adj, visit, inStack, safe)) {
                    return true;
                }
            }
            // Remove the node from the stack.
            inStack[node] = false;
            // If no cycle is found, mark the node as safe.
            safe[node] = true;
            return false;
        }
    
        // Find all eventual safe nodes in the graph
        vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
            int n = graph.size();
            vector<bool> visit(n), inStack(n), safe(n);
    
            for (int i = 0; i < n; i++) {
                dfs(i, graph, visit, inStack, safe);
            }
    
            vector<int> safeNodes;
            for (int i = 0; i < n; i++) {
                if (safe[i]) {
                    safeNodes.push_back(i);
                }
            }
    
            return safeNodes;
        }
    }
    /*
    Approach Summary:
    This solution uses DFS to detect cycles in the graph. It maintains three boolean vectors: `visit` to track visited nodes, `inStack` to detect cycles, and `safe` to mark nodes that do not lead to cycles. After performing DFS from each node, it collects nodes marked as safe.
    
    Complexity Analysis:
    - Time Complexity: O(N + E), where N is the number of nodes and E is the number of edges, because each node and edge is visited once.
    - Space Complexity: O(N), for the `visit`, `inStack`, and `safe` vectors.
    */
    

File: 801→900/802. Find Eventual Safe States/Solution.js
/**
 * Finds all eventual safe nodes in a graph.
 * @param {number[][]} graph
 * @return {number[]}
 */
var eventualSafeNodes = function (graph) {
  // Initialize an array to store safe nodes
  let results = [];

  // Use an object to keep track of visited nodes
  let visited = {};

  // Iterate over each node in the graph
  for (let i = 0; i < graph.length; i++) {
    // Check if the current node is safe
    if (isNodeSafe(graph, i, visited)) {
      // If safe, add it to the results array
      results.push(i);
    }
  }

  // Return the sorted array of safe nodes
  return results.sort((a, b) => a - b);
};

// Checks if a node is safe by performing DFS
function isNodeSafe(graph, node, visited) {
  // If the node is marked as visited (part of a cycle), it's not safe
  if (visited[node] === true) {
    return false;
  }

  // If the node is marked as safe, return true
  if (visited[node] === "safe") {
    return true;
  }

  // Mark the node as visited (part of the current path)
  visited[node] = true;

  // Check all neighbors of the node
  for (let i = 0; i < graph[node].length; i++) {
    // If any neighbor is not safe, the current node is not safe
    if (!isNodeSafe(graph, graph[node][i], visited)) {
      return false;
    }
  }

  // If all neighbors are safe, mark the node as safe
  visited[node] = "safe";

  return true;
}
/*
Approach Summary:
This solution uses DFS to identify safe nodes in a graph. A node is considered safe if it does not lead to a cycle. The `isNodeSafe` function checks each node's neighbors recursively and marks nodes as safe or part of a cycle based on the results.

Complexity Analysis:
- Time Complexity: O(N + E), where N is the number of nodes and E is the number of edges, because each node and edge is visited once.
- Space Complexity: O(N), for the `visited` object.
*/
